_posts/2020-10-02-producer-tuning.md

```diff
@@ -0,0 +1,307 @@
+
+```properties
+# ...
+acks=all
+# ...
+```
+Use the `acks=all` producer configuration in conjunction with the `min.insync.replicas` property for topics. You set the `min.insync.replicas` property in the `KafkaTopic` resource.
+
+The `min.insync.replicas` configuration sets the numbers of brokers that need to have logged a message before an acknowledgment is sent to the producer.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1alpha1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 3
+  config:
+    min.insync.replicas: 2
+    #...
+```
+Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.
+
+If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.
+
+Because of the additional checks, `acks=all` increases the latency between the producer sending a message and receiving acknowledgment. So you will have to consider the trade-off when investigating whether this is the right approach for you.
+
+### Ordered delivery
+
+You have two approaches to guaranteeing the order of message delivery from producers. And they depend, to a large degree, on whether or not you are using `acks=all` for data durability.
+
+If you are using `acks=all`, you can (and should) enable idempotence for the producer to ensure that messages are delivered only once.
+Using idempotence, IDs and sequence numbers are assigned to messages, so that the order is preserved even after a failed delivery.
+
+As idempotence preserves the message ordering, you can speed the process along by increasing the number of in-flight requests you allow at one time using the `max.in.flight.requests.per.connection`.
+
+Here we see an example configuration showing idempotence enabled, and used with `max.in.flight.requests.per.connection` and `acks=all`.
+
+```properties
+# ...
+enable.idempotence=true
+max.in.flight.requests.per.connection=5
+acks=all
+retries=2147483647
+```
+The `retries` property sets the number of retries when resending a failed message request. That impressive-looking number is the default and maximum value.
+
+> **Delivery timeout**
+>
+>If you are using `delivery.timeout.ms` in your producer configuration, producer requests will fail before the number of retries has been used if the timeout expires before a successful acknowledgment. The `delivery.timeout.ms` sets a limit on the time to wait for an acknowledgment of the success or failure to deliver a message. You can choose to leave `retries` unset and use `delivery.timeout.ms` to perform a similar function instead.
+
+There is a performance cost to introducing additional checks to the order of delivery.
+
+If you prefer not to use `acks=all` and idempotency,
+a lighter option is to set the number of in-flight requests to 1 (the default is 5) to preserve ordering.
```

Comment:
I wouldn't use "lighter", as the lightness really depends on the requirements. Not pipelining comes with its own costs.

Reply:
changed to _...another option is..._